### PR TITLE
Deduplicate reaction senders

### PR DESCRIPTION
**Pull request by jonnyandrew**
## Changes

Previously, if asked for the senders of any given reaction, `ReactionGroup` returned a vector containing possible duplicates.

After this change, the function returns only unique sender IDs.

## Context

According to [the spec](https://spec.matrix.org/v1.7/client-server-api/#annotations-client-behaviour), 

> **Annotations client behaviour**
> Multiple identical annotations (i.e., with the same event `type` and annotation `key`) from the same user (i.e., events with the same `sender`) should be treated as a single annotation.

I have taken this to mean that, when counting the number of reactions sent by a single user, the total should be either 1 or 0. And given there is no use case for retrieving senders in the timeline except to count them, the resulting list should contain only the unique sender IDs.

Note that we cannot exclude duplicate annotations altogether from a `ReactionGroup` as they must be known so that they can be redacted.

### CODE REVIEW

File: crates/matrix-sdk-ui/src/timeline/tests/reaction_group.rs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+use crate::timeline::{
+    tests::{ALICE, BOB},
+    ReactionGroup,
+};
+
+#[test]
+/// The Matrix spec does not allow duplicate annotations to be created but it
+/// is still possible for duplicates to be received over federation. And in
+/// that case, clients are expected to treat duplicates as a single annotation.
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/reaction_group.rs:24]:
```suggestion
/// The Matrix spec does not allow duplicate annotations to be created but it
/// is still possible for duplicates to be received over federation. And in
/// that case, clients are expected to treat duplicates as a single annotation.
#[test]
```

File: crates/matrix-sdk-ui/src/timeline/tests/reaction_group.rs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        group
+    };
+
+    assert_eq!(group.senders().collect::<Vec<_>>(), vec![&ALICE.to_owned(), &BOB.to_owned()]);
+}
+
+fn insert(group: &mut ReactionGroup, sender: &UserId, count: u64) {
+    for _ in 0..count {
+        let event_id = EventId::new(server_name!("dummy.server"));
+        group.0.insert((None::<OwnedTransactionId>, Some(event_id)), sender.to_owned());
```
Comment (jplatte) [inline crates/matrix-sdk-ui/src/timeline/tests/reaction_group.rs:39]: This should still compile, no?

```suggestion
        group.0.insert((None, Some(event_id)), sender.to_owned());
```

**Review by jplatte — APPROVED** [review]
This looks reasonable to me. I think we it makes sense for accesses to do the deduplication, we will need to store multiple txn IDs either way to, and still count the reaction of one of multiple identical ones is redacted.

File: crates/matrix-sdk-ui/src/timeline/tests/reaction_group.rs (hunk 2 above)
Comment (jonnyandrew) [reply crates/matrix-sdk-ui/src/timeline/tests/reaction_group.rs]: [`7da3215`](https://github.com/matrix-org/matrix-rust-sdk/pull/2132/commits/7da3215f5154a505e995dcde21b4468a8749917a)

**Pull request merged**